pytket_qir/generator.py

```diff
@@ -0,0 +1,407 @@
+            )
+            if (size := bit_reg.size) <= int_size:  # Widening by zero-padding.
+                bool_reg = list(map(bool, bit_reg_list)) + [False] * (int_size - size)
+            else:  # Narrowing by truncation.
+                bool_reg = list(map(bool, bit_reg_list[:int_size]))
+            ssa_var = cast(Value, self.module.builder.call(reg2var, [*bool_reg]))
+            self.ssa_vars[reg_name] = ssa_var
+            return ssa_var
+        else:
+            return cast(Value, self.ssa_vars[reg_name])
+
+    def _get_c_regs_from_com(self, command: Command) -> Tuple[List[str], List[str]]:
+        """Get classical registers from command op types."""
+        op = command.op
+        args = command.args
+        inputs: List[str] = []
+        outputs: List[str] = []
+
+        if isinstance(op, WASMOp):
+            for reglist, sizes in [
+                (inputs, op.input_widths),
+                (outputs, op.output_widths),
+            ]:
+                for in_width in sizes:
+                    assert in_width > 0
+                    com_bits = args[:in_width]
+                    args = args[in_width:]
+                    regname = com_bits[0].reg_name
+                    if com_bits != list(self.cregs[regname]):
+                        WASMError("WASM ops must act on entire registers.")
+                    reglist.append(regname)
+        elif isinstance(op, ClassicalExpBox):
+            for reglist, sizes in [
+                (
+                    inputs,
+                    list(
+                        map(
+                            lambda obj: obj.size
+                            if isinstance(obj, BitRegister)
+                            else None,
+                            op.get_exp().args,
+                        )
+                    ),
+                ),
+                (outputs, [op.get_n_o()]),
+            ]:
+                if not sizes:
+                    ClassicalExpBoxError(
+                        "ClassicalExpBox op input or output \
+                        registers have empty widths."
+                    )
+                for in_width in sizes:
+                    com_bits = args[:in_width]
+                    args = args[in_width:]
+                    regname = com_bits[0].reg_name
+                    if com_bits != list(self.cregs[regname]):
+                        ClassicalExpBoxError(
+                            "ClassicalExpBox ops must act on entire registers."
+                        )
+                    reglist.append(regname)
+        elif isinstance(op, SetBitsOp):
+            for reglist, sizes in [
+                (inputs, [op.n_inputs]),
+                (outputs, [op.n_outputs]),
+            ]:
+                for in_width in sizes:
+                    if in_width > 0:
+                        com_bits = args[:in_width]
+                        args = args[in_width:]
+                        regname = com_bits[0].reg_name
```

Comment:
Have you tried with latest pytket? This error was removed in https://github.com/CQCL/tket/pull/589 .

Reply:
Ok so now, it makes sense to throw an exception ? If `in_width  = 0` is valid `SetBits` that means the user is trying to assign a value to an empty register which very likely isn't what it intends to do.

Reply:
> Ok so now, it makes sense to throw an exception ? If `in_width = 0` is valid `SetBits` that means the user is trying to assign a value to an empty register which is very likely not what it intends to do.

I don't believe we should throw an exception just because it seems like an odd thing to do. By doing so, we may force users to insert special-case handling in their own code, which will be an annoyance for them. Edge cases should be handled painlessly, if possible.

Reply:
Yes but that can be the sign of a genuine bug ? I.e. an assignment to a wrong register ?

Reply:
(Remember that "the user" here could be some generated code.)
Basically, if it isn't easy to handle, fine, but I'd like to understand why -- since that would probably indicate some poor design choices upstream...

Reply:
Upstream where ? In the user process ? At the `SetBits` op level ?
Furthermore, I don't see a clear way to handle that case painlessly without further understanding of the user's intentions which looks like a complicated thing to do. For instance, setting the register size to an arbitrary value if empty ? I am not too sure it makes sense. But maybe I am wrong.

Reply:
> Upstream where ? In the user process ? At the `SetBits` op level ? Furthermore, I don't see a clear way to handle that case painlessly without further understanding of the user's intentions which looks like a complicated thing to do. For instance, setting the register size to an arbitrary value if empty ? I am not too sure it makes sense. But maybe I am wrong.

By "upstream" I mean in some code that this depends on: that may be elsewhere in pytket-qir or it may be in pytket or it may be in pyqir. But the user can reasonably expect to be able to do this, whether or not it seems sensible. The semantics of `SetBits` with no wires are unambiguous: it is a no-op. But sometimes code generation can emit no-ops and they shouldn't cause exceptions. I'm not sure why we need to understand the user's intentions beyond this? Perhaps I am missing something.

Reply:
Thanks for this. It is clear. What I meant is how do you disambiguate between a code generation that emits a no-op as part of a process and a genuine bug ? If you could, it would justify to handle the cases appropriately: ie raising an exception or painlessly going through. If not, how do you handle the case other than a generic behaviour like raising an exception ?

Reply:
> Thanks for this. It is clear. What I meant is how do you disambiguate between a code generation that emits a no-op as part of a process and a genuine bug ? If you could, it would justify to handle the cases appropriately: ie raising an exception or painlessly going through. If not, how do you handle the case other than a generic behaviour like raising an exception ?

We can't, and it's not our jpb to distinguish it -- any more than it is the job of gcc to error if a line of C code has no effect. We can optimize it away if we want -- but should allow it. If it's a bug, that's someone else's problem. Am I missing the point?

Reply:
No, I think we're on the same page. But in this context, since setting a `SetBits` with empty registers is now valid, it could still be a potential bug and I don't see any other way to handle the case without raising an exception if I am to strive for correctness ?